File: src/main/java/ru/serdtsev/homemoney/moneyoper/MoneyOperService.java
package ru.serdtsev.homemoney.moneyoper;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import ru.serdtsev.homemoney.account.Account;
import ru.serdtsev.homemoney.account.AccountRepository;
import ru.serdtsev.homemoney.account.Balance;
import ru.serdtsev.homemoney.balancesheet.BalanceSheet;
import ru.serdtsev.homemoney.balancesheet.BalanceSheetRepository;
import ru.serdtsev.homemoney.moneyoper.model.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.lang.String.format;
import static java.util.Objects.nonNull;
import static java.util.Objects.requireNonNull;
import static org.springframework.util.Assert.isTrue;

/**
 * Предоставляет методы работы с денежными операциями
 */
@Service
public class MoneyOperService {
  private static final String SEARCH_DATE_REGEX = "\\p{Digit}{4}-\\p{Digit}{2}-\\p{Digit}{2}";
  private static final String SEARCH_UUID_REGEX = "\\p{Alnum}{8}-\\p{Alnum}{4}-\\p{Alnum}{4}-\\p{Alnum}{4}-\\p{Alnum}{12}";
  private static final String SEARCH_MONEY_REGEX = "\\p{Digit}+\\.*\\p{Digit}*";
  private final Logger log = LoggerFactory.getLogger(this.getClass());
  private final BalanceSheetRepository balanceSheetRepo;
  private final MoneyOperRepo moneyOperRepo;
  private final RecurrenceOperRepo recurrenceOperRepo;
  private final AccountRepository accountRepo;
  private final LabelRepository labelRepo;

  @Autowired
  public MoneyOperService(BalanceSheetRepository balanceSheetRepo, MoneyOperRepo moneyOperRepo, RecurrenceOperRepo recurrenceOperRepo,
      AccountRepository accountRepo, LabelRepository labelRepo) {
    this.balanceSheetRepo = balanceSheetRepo;
    this.moneyOperRepo = moneyOperRepo;
    this.recurrenceOperRepo = recurrenceOperRepo;
    this.accountRepo = accountRepo;
    this.labelRepo = labelRepo;
  }

  public Optional<MoneyOper> findMoneyOper(UUID id) {
    return Optional.ofNullable(moneyOperRepo.findOne(id));
  }

  public void save(MoneyOper moneyOper) {
    moneyOperRepo.save(moneyOper);
  }

  public Optional<RecurrenceOper> findRecurrenceOper(UUID id) {
    return Optional.ofNullable(recurrenceOperRepo.findOne(id));
  }

  public void save(RecurrenceOper recurrenceOper) {
    recurrenceOperRepo.save(recurrenceOper);
  }

  /**
   * Возвращает следующие повторы операций.
   */
  Stream<MoneyOper> getNextRecurrenceOpers(BalanceSheet balanceSheet, String search, LocalDate beforeDate) {
    return getRecurrenceOpers(balanceSheet, search)
        .filter(recurrenceOper -> recurrenceOper.getNextDate().isBefore(beforeDate))
        .map(recurrenceOper -> createMoneyOperByTemplate(balanceSheet, recurrenceOper));
  }

  /**
   * Возвращает повторяющиеся операции.
   */
  public Stream<RecurrenceOper> getRecurrenceOpers(BalanceSheet balanceSheet, String search) {
    return recurrenceOperRepo.findByBalanceSheet(balanceSheet)
        .filter(recurrenceOper -> !recurrenceOper.getArc() && isOperMatchSearch(recurrenceOper, search));
  }

  /**
   * @return true, если шаблон операции соответствует строке поиска
   */
  private boolean isOperMatchSearch(RecurrenceOper recurrenceOper, String search) {
    MoneyOper template = recurrenceOper.getTemplate();
    if (StringUtils.isEmpty(search)) {
      return true;
    } else if (search.matches(SEARCH_DATE_REGEX)) {
      // по дате в формате ISO
      return recurrenceOper.getNextDate().isEqual(LocalDate.parse(search));
    } else if (search.matches(SEARCH_UUID_REGEX)) {
      // по идентификатору операции
      return template.getId().equals(UUID.fromString(search));
    } else if (search.matches(SEARCH_MONEY_REGEX)) {
      // по сумме операции
      return template.getItems()
          .stream()
          .anyMatch(change -> change.getValue().plus().compareTo(new BigDecimal(search)) == 0);
    } else {
      return template.getItems()
          .stream()
          .anyMatch(change -> change.getBalance().getName().toLowerCase().contains(search))
          || template.getComment().toLowerCase().contains(search)
          || template.getLabels().stream().anyMatch(label -> label.getName().toLowerCase().contains(search));
    }
  }

  private MoneyOper createMoneyOperByTemplate(BalanceSheet balanceSheet, RecurrenceOper recurrenceOper) {
    MoneyOper template = recurrenceOper.getTemplate();
    MoneyOper oper = new MoneyOper(UUID.randomUUID(), balanceSheet, MoneyOperStatus.recurrence,
        recurrenceOper.getNextDate(), 0, template.getLabels(), template.getComment(), template.getPeriod());
    oper.addItems(template.getItems());
    oper.setFromAccId(template.getFromAccId());
    oper.setAmount(template.getAmount());
    oper.setToAccId(template.getToAccId());
    oper.setToAmount(template.getToAmount());
    oper.setRecurrenceId(template.getRecurrenceId());
    return oper;
  }

  public void createRecurrenceOper(BalanceSheet balanceSheet, UUID operId) {
    MoneyOper sample = moneyOperRepo.findOne(operId);
    requireNonNull(sample);
    checkMoneyOperBelongsBalanceSheet(sample, balanceSheet.getId());
    MoneyOper template = newMoneyOper(balanceSheet, UUID.randomUUID(), MoneyOperStatus.template, null, null,
        sample.getLabels(), sample.getComment(), sample.getPeriod(), sample.getFromAccId(), sample.getToAccId(), sample.getAmount(),
        sample.getToAmount(), null, null);
    RecurrenceOper recurrenceOper = new RecurrenceOper(UUID.randomUUID(), balanceSheet, template, sample.getPerformed());
    recurrenceOper.skipNextDate();
    moneyOperRepo.save(template);
    recurrenceOperRepo.save(recurrenceOper);

    // todo поправить эту косоту
    template.setRecurrenceId(recurrenceOper.getId());
    moneyOperRepo.save(template);
    sample.setRecurrenceId(recurrenceOper.getId());
    moneyOperRepo.save(sample);
  }

  public void deleteRecurrenceOper(BalanceSheet balanceSheet, UUID recurrenceId) {
    RecurrenceOper recurrenceOper = recurrenceOperRepo.findOne(recurrenceId);
    recurrenceOper.arc();
    recurrenceOperRepo.save(recurrenceOper);
    log.info("RecurrenceOper '{}' moved to archive.", recurrenceId);
  }

  public void skipRecurrenceOper(BalanceSheet balanceSheet, UUID recurrenceId) {
    RecurrenceOper recurrenceOper = recurrenceOperRepo.findOne(recurrenceId);
    recurrenceOper.skipNextDate();
    recurrenceOperRepo.save(recurrenceOper);
  }

  /**
   * Создает экземпляр MoneyOper.
   */
  public MoneyOper newMoneyOper(BalanceSheet balanceSheet, UUID moneyOperId, MoneyOperStatus status, LocalDate performed,
      Integer dateNum, Collection<Label> labels, String comment, Period period, UUID fromAccId, UUID toAccId, BigDecimal amount,
      BigDecimal toAmount, UUID parentId, MoneyOper templateOper) {
    MoneyOper oper = new MoneyOper(moneyOperId, balanceSheet, status, performed, dateNum, labels, comment, period);
    if (nonNull(templateOper)) {
      oper.setRecurrenceId(templateOper.getRecurrenceId());
    }

    oper.setFromAccId(fromAccId);
    Account fromAcc = accountRepo.findOne(fromAccId);
    assert fromAcc != null;
    if (fromAcc instanceof Balance) {
      oper.addItem((Balance) fromAcc, amount.negate(), performed);
    }

    oper.setToAccId(toAccId);
    Account toAcc = accountRepo.findOne(toAccId);
    assert toAcc != null;
    if (toAcc instanceof  Balance) {
      oper.addItem((Balance) toAcc, toAmount, performed);
    }

    oper.setAmount(amount);
    oper.setToAmount(toAmount);

    if (parentId != null) {
      MoneyOper parentOper =  moneyOperRepo.findOne(parentId);
      assert parentOper != null;
      oper.setParentOper(parentOper);
    }

    return oper;
  }

  public void updateRecurrenceOper(BalanceSheet balanceSheet, RecurrenceOperDto recurrenceOperDto) {
    RecurrenceOper recurrenceOper = recurrenceOperRepo.findOne(recurrenceOperDto.getId());
    recurrenceOper.setNextDate(recurrenceOperDto.getNextDate());
    MoneyOper template = recurrenceOper.getTemplate();

    updateFromAccount(template, recurrenceOperDto.getFromAccId());
    updateToAccount(template, recurrenceOperDto.getToAccId());
    updateAmount(template, template.getAmount());
    updateToAmount(template, template.getToAmount());

    template.setComment(recurrenceOperDto.getComment());

    Collection<Label> labels = getLabelsByStrings(balanceSheet, recurrenceOperDto.getLabels());
    template.setLabels(labels);

    recurrenceOperRepo.save(recurrenceOper);
  }

  void updateAmount(MoneyOper oper, BigDecimal amount) {
    if (oper.getAmount().compareTo(amount) == 0) return;
    oper.getItems().stream()
        .filter(item -> item.getValue().signum() < 0)
        .forEach(item -> item.setValue(amount.negate()));
    oper.setAmount(amount);
  }

  void updateToAmount(MoneyOper oper, BigDecimal amount) {
    if (oper.getToAmount().compareTo(amount) == 0) return;
    oper.getItems().stream()
        .filter(item -> item.getValue().signum() > 0)
        .forEach(item -> item.setValue(amount));
    oper.setToAmount(amount);
  }

  void updateFromAccount(MoneyOper oper, UUID accId) {
    if (oper.getFromAccId().equals(accId)) return;
    replaceBalance(oper, oper.getFromAccId(), accId);
    oper.setFromAccId(accId);
  }

  private void replaceBalance(MoneyOper oper, UUID oldAccId, UUID newAccId) {
    oper.getItems().stream()
        .filter(item -> item.getBalance().getId().equals(oldAccId))
        .forEach(item -> {
          Balance balance = (Balance) accountRepo.findOne(newAccId);
          item.setBalance(balance);
        });
  }

  void updateToAccount(MoneyOper oper, UUID accId) {
    if (oper.getToAccId().equals(accId)) return;
    replaceBalance(oper, oper.getToAccId(), accId);
    oper.setToAccId(accId);
  }

  public void checkMoneyOperBelongsBalanceSheet(MoneyOper oper, UUID bsId) {
    isTrue(Objects.equals(oper.getBalanceSheet().getId(), bsId),
        format("MoneyOper id='%s' belongs the other balance sheet.", oper.getId()));
  }

  public List<Label> getLabelsByStrings(BalanceSheet balanceSheet, List<String> strLabels) {
    return strLabels
        .stream()
        .map(name -> findOrCreateLabel(balanceSheet, name))
        .collect(Collectors.toList());
  }

  Label findOrCreateLabel(BalanceSheet balanceSheet, String name) {
    Label label = labelRepo.findByBalanceSheetAndName(balanceSheet, name);
    return Optional.ofNullable(label)
        .orElseGet(() -> createSimpleLabel(balanceSheet, name));
  }

  private Label createSimpleLabel(BalanceSheet balanceSheet, String name) {
    Label label = new Label(UUID.randomUUID(), balanceSheet, name);
    labelRepo.save(label);
    return label;
  }

  List<String> getStringsByLabels(Collection<Label> labels) {
    return labels.stream()
        .map(Label::getName)
        .collect(Collectors.toList());
  }

  List<Label> getSuggestLabels(UUID bsId, MoneyOperDto moneyOper) {
    // Найдем 10 наиболее часто используемых меток-категорий за последние 30 дней.
    LocalDate startDate = LocalDate.now().minusDays(30);
    BalanceSheet balanceSheet = balanceSheetRepo.findOne(bsId);
    return moneyOperRepo.findByBalanceSheetAndStatusAndPerformedGreaterThan(balanceSheet, MoneyOperStatus.done, startDate)
      .flatMap(oper -> oper.getLabels().stream())
      .filter(label -> !moneyOper.getLabels().contains(label))
      .collect(Collectors.groupingBy(label -> label, Collectors.counting()))
      .entrySet()
      .stream()
      .sorted(Map.Entry.comparingByValue(Collections.reverseOrder()))
      .map(Map.Entry::getKey)
      .collect(Collectors.toList());
  }

  public MoneyOperDto moneyOperToDto(MoneyOper moneyOper) {
    MoneyOperDto moneyOperDto = new MoneyOperDto(moneyOper.getId(), moneyOper.getStatus(), moneyOper.getPerformed(), moneyOper.getFromAccId(),
        moneyOper.getToAccId(), moneyOper.getAmount().abs(), moneyOper.getCurrencyCode(),
        moneyOper.getToAmount(), moneyOper.getToCurrencyCode(), moneyOper.getPeriod(), moneyOper.getComment(),
        getStringsByLabels(moneyOper.getLabels()), moneyOper.getDateNum(), moneyOper.getParentOperId(),
        moneyOper.getRecurrenceId(), moneyOper.getCreated());
    moneyOperDto.setFromAccName(getAccountName(moneyOper.getFromAccId()));
    moneyOperDto.setToAccName(getAccountName(moneyOper.getToAccId()));
    moneyOperDto.setType(moneyOper.getType().name());
    moneyOperDto.setItems(moneyOper.getItems());
    return moneyOperDto;
  }

  public String getAccountName(UUID accountId) {
    Account account = accountRepo.findOne(accountId);
    return account.getName();
  }

  Stream<Label> getLabels(UUID bsId) {
    BalanceSheet balanceSheet = balanceSheetRepo.findOne(bsId);
    return labelRepo.findByBalanceSheetOrderByName(balanceSheet);
  }
}

File: src/main/kotlin/ru/serdtsev/homemoney/moneyoper/MoneyOperService.kt
package ru.serdtsev.homemoney.moneyoper

import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.core.convert.ConversionService
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import ru.serdtsev.homemoney.account.AccountRepository
import ru.serdtsev.homemoney.account.BalanceRepository
import ru.serdtsev.homemoney.account.model.Reserve
import ru.serdtsev.homemoney.account.model.ServiceAccount
import ru.serdtsev.homemoney.balancesheet.BalanceSheet
import ru.serdtsev.homemoney.balancesheet.BalanceSheetRepository
import ru.serdtsev.homemoney.moneyoper.model.*
import java.math.BigDecimal
import java.time.LocalDate
import java.util.*

/**
 * Предоставляет методы работы с денежными операциями
 */
@Service
class MoneyOperService @Autowired constructor(
        @Qualifier("conversionService") private val conversionService: ConversionService,
        private val balanceSheetRepo: BalanceSheetRepository,
        private val moneyOperRepo: MoneyOperRepo,
        private val recurrenceOperRepo: RecurrenceOperRepo,
        private val accountRepo: AccountRepository,
        private val balanceRepo: BalanceRepository,
        private val labelRepo: LabelRepository
) {
    fun save(moneyOper: MoneyOper) {
        moneyOperRepo.save(moneyOper)
    }

    fun findRecurrenceOper(id: UUID): RecurrenceOper? {
        return recurrenceOperRepo.findByIdOrNull(id)
    }

    fun save(recurrenceOper: RecurrenceOper) {
        recurrenceOperRepo.save(recurrenceOper)
    }

    /**
     * Возвращает следующие повторы операций.
     */
    fun getNextRecurrenceOpers(balanceSheet: BalanceSheet, search: String, beforeDate: LocalDate?): List<MoneyOper> {
        return getRecurrenceOpers(balanceSheet, search)
                .filter { it.nextDate.isBefore(beforeDate) }
                .map { createMoneyOperByRecurrenceOper(balanceSheet, it) }
    }

    /**
     * Возвращает повторяющиеся операции.
     */
    fun getRecurrenceOpers(balanceSheet: BalanceSheet, search: String): List<RecurrenceOper> =
            recurrenceOperRepo.findByBalanceSheet(balanceSheet).filter { !it.arc && isOperMatchSearch(it, search) }

    /**
     * @return true, если шаблон операции соответствует строке поиска
     */
    private fun isOperMatchSearch(recurrenceOper: RecurrenceOper, search: String): Boolean {
        val template = recurrenceOper.template
        return when {
            search.isBlank() -> true
            search.matches(SearchDateRegex) -> {
                // по дате в формате ISO
                recurrenceOper.nextDate.isEqual(LocalDate.parse(search))
            }
            search.matches(SearchUuidRegex) -> {
                // по идентификатору операции
                template.id == UUID.fromString(search)
            }
            search.matches(SearchMoneyRegex) -> {
                // по сумме операции
                template.items.any { it.value.plus().compareTo(BigDecimal(search)) == 0 }
            }
            else -> {
                (template.items.any { it.balance.name.toLowerCase().contains(search) }
                        || template.labels.any { it.name.toLowerCase().contains(search) })
                        || template.comment?.toLowerCase()?.contains(search) ?: false
            }
        }
    }

    private fun createMoneyOperByRecurrenceOper(balanceSheet: BalanceSheet, recurrenceOper: RecurrenceOper): MoneyOper {
        val template = recurrenceOper.template
        val performed = recurrenceOper.nextDate
        val oper = MoneyOper(balanceSheet, MoneyOperStatus.recurrence, performed, 0, template.labels,
                template.comment, template.period)
        template.items.forEach { oper.addItem(it.balance, it.value, performed) }
        oper.recurrenceId = template.recurrenceId
        return oper
    }

    fun createRecurrenceOper(balanceSheet: BalanceSheet, operId: UUID) {
        val sample = moneyOperRepo.findByIdOrNull(operId)!!
        checkMoneyOperBelongsBalanceSheet(sample, balanceSheet.id)
        val template = MoneyOper(balanceSheet, MoneyOperStatus.template, sample.performed, 0, sample.labels,
                sample.comment, sample.period)
        sample.items.forEach { template.addItem(it.balance, it.value, it.performed) }
        val recurrenceOper = RecurrenceOper(balanceSheet, template, sample.performed)
        recurrenceOper.skipNextDate()
        moneyOperRepo.save(template)
        recurrenceOperRepo.save(recurrenceOper)

        // todo поправить эту косоту
        template.recurrenceId = recurrenceOper.id
        moneyOperRepo.save(template)
        sample.recurrenceId = recurrenceOper.id
        moneyOperRepo.save(sample)
    }

    fun deleteRecurrenceOper(balanceSheet: BalanceSheet, recurrenceId: UUID) {
        val recurrenceOper = recurrenceOperRepo.findByIdOrNull(recurrenceId)!!
        recurrenceOper.arc()
        recurrenceOperRepo.save(recurrenceOper)
        log.info("RecurrenceOper '{}' moved to archive.", recurrenceId)
    }

    fun skipRecurrenceOper(balanceSheet: BalanceSheet, recurrenceId: UUID) {
        val recurrenceOper = recurrenceOperRepo.findByIdOrNull(recurrenceId)!!
        recurrenceOper.skipNextDate()
        recurrenceOperRepo.save(recurrenceOper)
    }

    fun updateRecurrenceOper(balanceSheet: BalanceSheet, recurrenceOperDto: RecurrenceOperDto) {
        val origRecurrenceOper = recurrenceOperRepo.findByIdOrNull(recurrenceOperDto.id)!!
        origRecurrenceOper.nextDate = recurrenceOperDto.nextDate
        val origTemplate = origRecurrenceOper.template
        recurrenceOperDto.items.forEach { item ->
            val origItem = origTemplate.items.firstOrNull { origItem -> origItem.id == item.id }
            if (origItem != null) with (origItem) {
                balance = balanceRepo.findByIdOrNull(item.balanceId)!!
                value = item.value
                index = item.index
            } else {
                val balance = balanceRepo.findByIdOrNull(item.balanceId)!!
                origTemplate.addItem(balance, item.value, index = item.index)
            }
        }
        origTemplate.items.removeIf {
            origTemplateItem -> !recurrenceOperDto.items.any { it.id == origTemplateItem.id }
        }
        origTemplate.comment = recurrenceOperDto.comment
        val labels: Collection<Label> = getLabelsByStrings(balanceSheet, recurrenceOperDto.labels)
        origTemplate.setLabels(labels)
        recurrenceOperRepo.save(origRecurrenceOper)
    }

    fun moneyOperDtoToMoneyOper(balanceSheet: BalanceSheet, moneyOperDto: MoneyOperDto): MoneyOper {
        val dateNum = moneyOperDto.dateNum ?: 0
        val labels = getLabelsByStrings(balanceSheet, moneyOperDto.labels)
        val period = moneyOperDto.period ?: Period.month
        val oper = MoneyOper(balanceSheet, MoneyOperStatus.pending, moneyOperDto.operDate, dateNum, labels,
                moneyOperDto.comment, period)
        moneyOperDto.items.forEach {
            val balance = balanceRepo.findByIdOrNull(it.balanceId)!!
            oper.addItem(balance, it.value, it.performedAt, it.index)
        }
        return oper
    }

    fun updateMoneyOper(toOper: MoneyOper, fromOper: MoneyOper) {
        fromOper.items.forEach { item ->
            val origItem = toOper.items.firstOrNull { origItem -> origItem.id == item.id }
            if (origItem != null) with (origItem) {
                performed = item.performed
                balance = item.balance
                value = item.value
                index = item.index
            } else {
                toOper.addItem(item.balance, item.value, item.performed, item.index)
            }
        }
        toOper.items.retainAll(fromOper.items)
        toOper.performed = fromOper.performed
        toOper.setLabels(fromOper.labels)
        toOper.dateNum = fromOper.dateNum
        toOper.period = fromOper.period
        toOper.comment = fromOper.comment
    }

    fun checkMoneyOperBelongsBalanceSheet(oper: MoneyOper, bsId: UUID) =
            assert(oper.balanceSheet.id == bsId) { "MoneyOper id='${oper.id}' belongs the other balance sheet." }

    fun getLabelsByStrings(balanceSheet: BalanceSheet, strLabels: List<String>): MutableList<Label> =
            strLabels.map { findOrCreateLabel(balanceSheet, it) }.toMutableList()

    fun findOrCreateLabel(balanceSheet: BalanceSheet, name: String): Label =
            labelRepo.findByBalanceSheetAndName(balanceSheet, name) ?: run { createSimpleLabel(balanceSheet, name) }

    private fun createSimpleLabel(balanceSheet: BalanceSheet, name: String): Label =
            Label(UUID.randomUUID(), balanceSheet, name).apply { labelRepo.save(this) }

    fun getStringsByLabels(labels: Collection<Label>): List<String> = labels.map { it.name }

    fun getSuggestLabels(bsId: UUID, moneyOper: MoneyOperDto): List<Label> {
        // Найдем 10 наиболее часто используемых меток-категорий за последние 30 дней.
        val startDate = LocalDate.now().minusDays(30)
        val balanceSheet = balanceSheetRepo.findById(bsId).get()
        return moneyOperRepo.findByBalanceSheetAndStatusAndPerformedGreaterThan(balanceSheet, MoneyOperStatus.done, startDate)
                .flatMap { it.labels }
                .filter { !moneyOper.labels.contains(it.name) }
                .groupingBy { it }.eachCount()
                .entries
                .sortedByDescending { it.value }
                .map { it.key }
    }

    fun moneyOperToDto(moneyOper: MoneyOper): MoneyOperDto =
            MoneyOperDto(moneyOper.id, moneyOper.status, moneyOper.performed, moneyOper.period, moneyOper.comment,
                    getStringsByLabels(moneyOper.labels), moneyOper.dateNum, moneyOper.getParentOperId(),
                    moneyOper.recurrenceId, moneyOper.created).apply {
                if (moneyOper.type == MoneyOperType.transfer && moneyOper.items.any { it.balance is Reserve }) {
                    val operItem = moneyOper.items.first { it.balance is Reserve }
                    type = if (operItem.value.signum() > 0) MoneyOperType.income.name else MoneyOperType.expense.name
                } else
                    type = moneyOper.type.name
                items = moneyOper.items
                        .map { conversionService.convert(it, MoneyOperItemDto::class.java)!! }
                        .sortedBy { it.value }
            }

    fun getAccountName(accountId: UUID): String {
        val account = accountRepo.findByIdOrNull(accountId)!!
        return account.name
    }

    fun getLabels(bsId: UUID): List<Label> {
        val balanceSheet = balanceSheetRepo.findByIdOrNull(bsId)!!
        return labelRepo.findByBalanceSheetOrderByName(balanceSheet)
    }

    companion object {
        private val log = KotlinLogging.logger {  }
        private val SearchDateRegex = "\\p{Digit}{4}-\\p{Digit}{2}-\\p{Digit}{2}".toRegex()
        private val SearchUuidRegex = "\\p{Alnum}{8}-\\p{Alnum}{4}-\\p{Alnum}{4}-\\p{Alnum}{4}-\\p{Alnum}{12}".toRegex()
        private val SearchMoneyRegex = "\\p{Digit}+\\.*\\p{Digit}*".toRegex()
    }
}